src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+
+    /// Create a new `BatchBuilder` with optional description of the batch.
+    pub fn new(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder {
+            description: Some(description.into()),
+            ..Self::create_empty()
+        }
+    }
+
+    fn success(&mut self, success_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.success = Some(success_cb.into());
+        self
+    }
+
+    fn complete(&mut self, complete_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.complete = Some(complete_cb.into());
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(&mut self, success_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with only `complete` callback specified.
+    pub fn with_complete_callback(&mut self, complete_cb: Job) -> Batch {
+        self.complete(complete_cb);
+        self.success(None);
+        self.build()
+    }
+
+    /// Create a `Batch` with both `success` and `complete` callbacks specified.
+    pub fn with_callbacks(&mut self, success_cb: Job, complete_cb: Job) -> Batch {
+        self.success(success_cb);
+        self.complete(complete_cb);
+        self.build()
+    }
+}
+
+pub struct BatchHandle<'a, S: Read + Write> {
+    bid: String,
+    prod: &'a mut Producer<S>,
+}
+
+impl<'a, S: Read + Write> BatchHandle<'a, S> {
+    /// ID issued by the Faktory server to this batch.
+    pub fn id(&self) -> &str {
+        self.bid.as_ref()
+    }
+
+    pub(crate) fn new(bid: String, prod: &mut Producer<S>) -> BatchHandle<'_, S> {
+        BatchHandle { bid, prod }
+    }
+
+    /// Add the given job to the batch.
+    pub fn add(&mut self, mut job: Job) -> Result<(), Error> {
+        job.custom.insert("bid".into(), self.bid.clone().into());
+        self.prod.enqueue(job)
+    }
+
+    /// Initiate a child batch of jobs.
+    pub fn start_batch(&mut self, mut batch: Batch) -> Result<BatchHandle<'_, S>, Error> {
+        batch.parent_bid = Some(self.bid.clone());
+        self.prod.start_batch(batch)
+    }
+
+    /// Commit this batch.
+    ///
+    /// The Faktory server will not queue any callbacks, unless the batch is committed.
+    /// Committing an empty batch will make the server queue the callback(s) right away.
+    /// Once committed, the batch can still be re-opened with [open_batch](struct.Producer.html#method.open_batch),
+    /// and extra jobs can be added to it.
+    pub fn commit(self) -> Result<(), Error> {
+        self.prod.commit_batch(self.bid)
+    }
+}
+
+/// Batch status retrieved from Faktory server.
+#[derive(Deserialize, Debug)]
+pub struct BatchStatus {
+    // Fields "bid", "created_at", "description", "total", "pending", and "failed"
+    // are described in the docs: https://github.com/contribsys/faktory/wiki/Ent-Batches#status
+    /// Id of this batch.
+    pub bid: String,
```

Comment:
Might be cool to have a `BatchStatus::open` method that opens this `bid` as a handy shortcut?

Reply:
Added this method. So they can go with:
```
let mut t = Tracker::connect(Some(&url)).unwrap();
let status = t.get_batch_status(bid.clone()).unwrap().unwrap();
let mut b = status.open(&mut p).unwrap().unwrap(); // using the bid from the retrieved BatchStatus
b.add(extra_job).unwrap();
```